core/src/main/java/org/apache/calcite/sql2rel/StandardConvertletTable.java

```diff
@@ -308,12 +308,20 @@ private static RexNode convertNvl(SqlRexContext cx, SqlCall call) {
         cx.convertExpression(call.getOperandList().get(1));
     final RelDataType type =
         cx.getValidator().getValidatedNodeType(call);
+    // Preserve Operand Nullability
     return rexBuilder.makeCall(type, SqlStdOperatorTable.CASE,
         ImmutableList.of(
             rexBuilder.makeCall(SqlStdOperatorTable.IS_NOT_NULL,
                 operand0),
-            rexBuilder.makeCast(type, operand0),
-            rexBuilder.makeCast(type, operand1)));
+            rexBuilder.makeCast(
+                cx.getTypeFactory()
+                    .createTypeWithNullability(type, operand0.getType().isNullable()),
+                operand0),
+            rexBuilder.makeCast(
+                cx.getTypeFactory()
+                    .createTypeWithNullability(type, operand1.getType().isNullable()),
+                operand1)
+        ));
   }
```

Comment:
Could you add a test case, @jaystarshot ?

Reply:
@chunweilei  There is no method to test the Operands of an Operator in the test framework, I think I need to create a new one and/or test it as an integration test.